### PR TITLE
froide: init at 0-unstable-2024-11-22

### PR DESCRIPTION
**Pull request by onny**
Adding web app [froide](https://github.com/okfde/froide), a portal for creating and following freedom of information act requests. Used by [FragDenStaat](https://fragdenstaat.de) project.

Adding it only as Python module as such it can also be used by other projects - in my use case https://github.com/NixOS/nixpkgs/pull/349750

Related:
  - https://github.com/NixOS/nixpkgs/pull/355390
  - https://github.com/okfde/froide/issues/893
  - https://github.com/okfde/froide/issues/902
  - https://github.com/okfde/froide/issues/901

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/web-apps/froide.nix [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+}:
+let
+
+  cfg = config.services.froide;
+  format = pkgs.formats.toml { };
+
+in {
+  options.services.froide = {
+
+    enable = lib.mkEnableOption "Gouvernment planer web app Govplan";
```
Comment (pluiedev) [inline nixos/modules/services/web-apps/froide.nix:15]: User-facing documentation should be (unfortunately) in English

File: nixos/modules/services/web-apps/froide.nix [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        [
+          ''
+            ${lib.getExe' config.services.postgresql.package "psql"} -d froide -f "${sqlFile}"
+          ''
+        ];
+
+      froide = {
+        description = "Gouvernment planer Govplan";
+        serviceConfig = {
+          ExecStart = "${pkgs.froide}/bin/froide runserver 0.0.0.0:8000";
```
Comment (pluiedev) [inline nixos/modules/services/web-apps/froide.nix:86]: Consider adding host & port options

File: nixos/modules/services/web-apps/froide.nix [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+          WorkingDirectory = "/var/lib/froide";
+          StateDirectory = [ "froide" ];
+          DynamicUser = true;
+          EnvironmentFile = [
+            ( format.generate "froide-env" cfg.settings )
+            cfg.secrets
+          ];
```
Comment (pluiedev) [inline nixos/modules/services/web-apps/froide.nix:93]: Optional: hardening?

File: nixos/tests/froide.nix [hunk 4]
```diff
@@ -0,0 +1,90 @@
+import ./make-test-python.nix ({ lib, ... }: {
+  name = "paperless";
```
Comment (pluiedev) [inline nixos/tests/froide.nix:2]: This is still paperless

File: pkgs/by-name/fr/froide/package.nix [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+  };
+
+  postBuild = ''
+    pnpm run build
+  '';
+
+  postInstall = ''
+    cp manage.py $out/
+    makeWrapper $out/manage.py $out/bin/froide \
+      --prefix PYTHONPATH : "$PYTHONPATH" \
```
Comment (pluiedev) [inline pkgs/by-name/fr/froide/package.nix:115]: This should already happen when patching interpreters in fixupPhase

File: pkgs/by-name/fr/froide/package.nix [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+  ];
+
+  checkInputs = with python.pkgs; [
+    pytest-factoryboy
+    time-machine
+    beautifulsoup4
+  ];
+
+  env = {
+    PGDATABASE = "froide";
```
Comment (pluiedev) [inline pkgs/by-name/fr/froide/package.nix:136]: Do these need to be specified at build time?

File: pkgs/by-name/fr/froide/package.nix [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+  };
+
+  # Need to enable postgis on test server
+  doCheck = false;
+
+  meta = {
+    description = "Freedom of Information Portal";
+    homepage = "https://github.com/okfde/froide";
+    license = lib.licenses.mit;
+    maintainers = [ lib.maintainers.onny ];
```
Comment (pluiedev) [inline pkgs/by-name/fr/froide/package.nix:158]: Missing `mainProgram`, and `platforms` if applicable

File: pkgs/development/python-modules/django-fsm/default.nix [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    rev = "refs/tags/${version}";
+    hash = "sha256-woN0F4hTaPk8HTGNT6zQlZDJ9SCVRut9maKSlDmalUE=";
+  };
+
+  build-system = [ setuptools ];
+
+  dependencies = [ django ];
+
+  checkPhase = ''
+    ${python.interpreter} manage.py test
```
Comment (pluiedev) [inline pkgs/development/python-modules/django-fsm/default.nix:29]: Shouldn't `python3` already be in PATH? Plus, python means Python 2 (!)

**Review by pluiedev — CHANGES_REQUESTED** [review]
Maybe it would be better to clarify which country/region this app is for? Also please run nixfmt on your changes (except for large files like all-packages.nix, python-modules.nix, etc), and reorder commits so that froide's dependencies are added before froide itself

File: pkgs/development/python-modules/froide/default.nix [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+          license = lib.licenses.mit;
+          knownVulnerabilities = [ "Package is marked as discontinued upstream." ];
+          maintainers = [ lib.maintainers.onny ];
+        };
+      };
+    };
+  };
+
+in
+python.pkgs.buildPythonPackage rec {
```
Comment (pluiedev) [inline pkgs/development/python-modules/froide/default.nix:57]: This is supposed to be used as an application, right? Then it shouldn't be under `development/python-modules` and should be built with `buildPythonApplication`

File: pkgs/development/python-modules/froide/default.nix [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+        pythonImportsCheck = [ "django_fsm" ];
+
+        meta = {
+          description = "Django friendly finite state machine support";
+          homepage = "https://github.com/viewflow/django-fsm";
+          license = lib.licenses.mit;
+          knownVulnerabilities = [ "Package is marked as discontinued upstream." ];
+          maintainers = [ lib.maintainers.onny ];
+        };
+      };
```
Comment (pluiedev) [inline pkgs/development/python-modules/froide/default.nix:52]: Not necessary — just add it as another regular Python module and drop it when upstream fixes it

**Review by pluiedev — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/froide/default.nix (hunk 9 above)
Comment (onny) [reply pkgs/development/python-modules/froide/default.nix:57]: For me I don't want it to use as a standalone application (too much effort to package). I only want to use it as module which is also possible. Using it as a module for this application https://github.com/NixOS/nixpkgs/pull/349750

**Comment by onny** [comment]
Thank you @pluiedev fixed it :)

**Review by pluiedev — APPROVED** [review]
LGTM though I'm not a Python domain expert. Would want to get feedback from them first

File: pkgs/by-name/fr/froide/package.nix [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+
+  postBuild = ''
+    pnpm run build
+  '';
+
+  postInstall = ''
+    cp -r build manage.py $out/${python.sitePackages}/froide/
+    makeWrapper $out/${python.sitePackages}/froide/manage.py $out/bin/froide \
+      --prefix PYTHONPATH : "$PYTHONPATH" \
+      --set GDAL_LIBRARY_PATH "${gdal}/lib/libgdal.so" \
```
Comment (dotlambda) [inline pkgs/by-name/fr/froide/package.nix]: Should this use
```suggestion
      --set GDAL_LIBRARY_PATH "${gdalMinimal}/lib/libgdal.so" \
```
just like postgis?

File: pkgs/top-level/python-packages.nix [hunk 12]
```diff
@@ -4888,6 +4890,8 @@ self: super: with self; {
 
   fritzconnection = callPackage ../development/python-modules/fritzconnection { };
 
+  froide = toPythonModule (pkgs.froide.override { python3Packages = self; });
```
Comment (dotlambda) [inline pkgs/top-level/python-packages.nix]: Why??? Removing in https://github.com/NixOS/nixpkgs/pull/468487.

**Pull request merged**